2000-08-25  Jason Merrill  <jason@redhat.com>

	* integrate.c (expand_inline_function): Pull out the original decl.

2001-04-05  Jakub Jelinek  <jakub@redhat.com>

	* gcc.dg/20010405-1.c: New test.

--- gcc/testsuite/gcc.dg/20010405-1.c.jj	Thu Apr  5 17:04:38 2001
+++ gcc/testsuite/gcc.dg/20010405-1.c	Thu Apr  5 16:53:50 2001
@@ -0,0 +1,19 @@
+__inline__ double bar(double *x)
+{
+  static double a;
+  a = *x >= 0 ? *x : -*x;
+  return a;
+}
+
+int main(void)
+{
+  extern double bar();
+  double a;
+  static double b;
+  int r;
+  for (r = 1; r < 3; r++) {
+    a = 1.0;
+    b = bar(&a);
+  }
+  return 0;
+}
--- gcc/integrate.c.jj	Wed Apr  4 19:07:10 2001
+++ gcc/integrate.c	Thu Apr  5 21:08:03 2001
@@ -653,6 +653,10 @@ expand_inline_function (fndecl, parms, t
   if (max_regno < FIRST_PSEUDO_REGISTER)
     abort ();
 
+  /* Pull out the decl for the function definition; fndecl may be a
+     local declaration, which would break DECL_ABSTRACT_ORIGIN.  */
+  fndecl = inl_f->decl;
+
   nargs = list_length (DECL_ARGUMENTS (fndecl));
 
   if (cfun->preferred_stack_boundary < inl_f->preferred_stack_boundary)
